src/sound_len.rs

```diff
@@ -0,0 +1,150 @@
+        enable_gapless: true,
+        ..Default::default()
+    };
+
+    let probed = match get_probe().format(&hint, mss, &fmt_opts, &meta_opts) {
+        Ok(r) => r,
+        Err(e) => return Err(SoundLen(format!("Probe error: {}", e))),
+    };
+
+    match sound_length_simple(&probed) {
+        Ok(r) => return Ok(format!("{:.3}", r as f32)),
+        Err(_e) => (),
+    };
+
+    match sound_length_decode(probed) {
+        Ok(r) => Ok(format!("{:.3}", r as f32)),
+        Err(e) => Err(e),
+    }
+}
+
+fn sound_length_simple(probed: &ProbeResult) -> Result<f64> {
+    let format = &probed.format;
+
+    let track = match format.default_track() {
+        Some(r) => r,
+        None => return Err(SoundLen("Could not get default track".to_string())),
+    };
+
+    let time_base = match track.codec_params.time_base {
+        Some(r) => r,
+        None => return Err(SoundLen("Codec does not provide a time base.".to_string())),
+    };
+
+    let n_frames = match track.codec_params.n_frames {
+        Some(r) => r,
+        None => return Err(SoundLen("Codec does not provide frame count".to_string())),
+    };
+
+    let time = time_base.calc_time(n_frames);
+    let duration = Duration::from_secs(time.seconds) + Duration::from_secs_f64(time.frac);
+
+    Ok(duration.as_secs_f64() * 10.0)
+}
+
+fn sound_length_decode(probed: ProbeResult) -> Result<f64> {
+    let mut format = probed.format;
+
+    let track = match format.default_track() {
+        Some(r) => r,
+        None => return Err(SoundLen("Could not get default track".to_string())),
+    };
+
+    // Grab the number of frames of the track
+    let samples_capacity = if let Some(n_frames) = track.codec_params.n_frames {
+        n_frames as f64
+    } else {
+        0.0
+    };
+
+    // Create a decoder using the provided codec parameters in the track.
+    let decoder_opts: DecoderOptions = Default::default();
+    let mut decoder = match get_codecs().make(&track.codec_params, &decoder_opts) {
+        Ok(r) => r,
+        Err(e) => return Err(SoundLen(format!("Decoder creation error: {}", e))),
+    };
+
+    // Try to grab a data packet from the container
+    let encoded_packet = match format.next_packet() {
+        Ok(r) => r,
+        Err(e) => return Err(SoundLen(format!("Next_packet error: {}", e))),
+    };
+
+    // Try to decode the data packet
+    let decoded_packet = match decoder.decode(&encoded_packet) {
+        Ok(r) => r,
+        Err(e) => return Err(SoundLen(format!("Decode error: {}", e))),
+    };
+
+    // Grab the sample rate from the spec of the buffer.
+    let sample_rate = decoded_packet.spec().rate as f64;
+    // Math!
+    let duration_in_desciseconds = samples_capacity / sample_rate * 10.0;
+    Ok(duration_in_desciseconds)
+}
+
+byond_fn!(
+    fn sound_len_list(list) {
+        Some(get_sound_length_list(list))
+    }
+);
+
+fn get_sound_length_list(list: &str) -> String {
```

Comment:
do you really need all of the explicit types inside this function? pretty sure they can be figured out by the compiler fine